xdsl/transforms/lower_csl_stencil.py

```diff
@@ -219,6 +235,77 @@ def match_and_rewrite(self, op: csl_stencil.ApplyOp, rewriter: PatternRewriter,
         )
 
 
+@dataclass(frozen=True)
+class GenerateCoeffAPICalls(RewritePattern):
+    """
+    Generates calls to the stencil_comms API to set coefficients.
+
+    The API currently supports only f32 coeffs.
+
+    Todo:
+      * reset coeffs for any subsequent apply op that does not generate a `setCoeffs` API call
+      * check if coeffs need to be set repeatedly (in loops or for multiple applies)
+      * hoist API call for loops with exactly one apply op
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: csl_stencil.ApplyOp, rewriter: PatternRewriter, /):
+        if not (wrapper := _get_module_wrapper(op)) or op.coeffs is None:
+            return
+        coeffs = list(op.coeffs)
+        elem_t = coeffs[0].coeff.type
+        pattern = wrapper.get_param_value("pattern").value.data
+        neighbours = pattern - 1
+        cmap: dict[csl.Direction, list[AnyFloatAttr]] = {
+            csl.Direction.NORTH: [FloatAttr(f, elem_t) for f in [0] + neighbours * [1]],
+            csl.Direction.SOUTH: [FloatAttr(f, elem_t) for f in [0] + neighbours * [1]],
+            csl.Direction.EAST: [FloatAttr(f, elem_t) for f in [0] + neighbours * [1]],
+            csl.Direction.WEST: [FloatAttr(f, elem_t) for f in [0] + neighbours * [1]],
```

Comment:
If these are all the same, maybe a single array with `.copy()` would be better? To avoid copy-pasting

Reply:
done as suggested